s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -65,6 +65,24 @@ private[alpakka] final class S3Stream(credentials: AWSCredentials,
     Source.fromFuture(request(s3Location, range).flatMap(entityForSuccess).map(_.dataBytes)).flatMapConcat(identity)
   }
 
+  def listBucket(bucket: String, prefix: Option[String] = None): Source[String, NotUsed] = {
+    def listBucketCall(continuation_token: Option[String] = None): Future[ListBucketResult] =
+      signAndGetAs[ListBucketResult](HttpRequests.listBucket(bucket, region, prefix, continuation_token))
+
+    def fileSourceFromFuture(f: Future[ListBucketResult]): Source[String, NotUsed] =
+      Source
+        .fromFuture(f)
+        .flatMapConcat((res: ListBucketResult) => {
```

Comment:
After digging a little here, Im not quite sure how you use this technique to page through the results properly. I dont how statefulMapConcat gives you access to the previous value than continues to page until completed.